docs/embedded_builds.md

```diff
@@ -0,0 +1,190 @@
+# Embedding dependencies
+
+To make bpftrace more portable, it has long supported an alpine-based musl
+build, which statically compiled bpftrace resulting in no runtime linking
+required.
+
+The drawback to this approach is that LLVM libraries, even when statically
+compiled, depends on symbols from libdl, and works best and most predictably
+when dynamically linked to libc.
+
+To embed everything except for libc, building LLVM and Clang from source is
+supported. This allows for linking to arbitrary libc targets dynamically, which
+may provide the best of both worlds between a purely static and a purely
+dynamically-linked bpftrace executable.
+
+For this reason, there is CMake support in the bpftrace project to build LLVM
+and Clang from source, as these are the heaviest dependencies of bpftrace.
+Other library dependencies can be obtained by most package managers reliably.
+
+Upstream packages provided by package maintainers can't be depended on to
+present all of the necessary built artifacts to statically link against LLVM and
+Clang, despite the project [LLVM catering to this with its own CMake build
+process](https://llvm.org/docs/CMake.html#embedding-llvm-in-your-project).
+
+# Configuration flags
+
+To make an semi-static executable that includes everything except libc, the
+CMake option `STATIC_LIBC:BOOL=OFF` can be added. This should allow for a
+bpftrace executable that can run on any system with a compatible libc.
+
+To assist in this, new configuration flags allow to download, configure, build
+and embed static libraries for all dependencies of bpftrace, rather than
+relying on system libraries.
+
+To build the necessary static LLVM or Clang libraries, `-DEMBED_LLVM:BOOL=ON`
+and `-DEMBED_CLANG=ON` can be used respectively. Clang can link to system LLVM
+as well, but may need to be patched.
+
+It is possible to link bpftrace's embedded libraries with system libraries, so
+linking to a distribution specific LLVM is possible. Many distributions do not
+include `libclang.a`,  using `-DEMBED_LIBCLANG_ONLY=ON` will build only
+`libclang.a`, allowing for linking to system LLVM and Clang libraries on Debian,
+as opposed to Vanilla LLVM sources.
+
+# Build times
+
+To build from scratch on a modern development laptop with an 8-way parallel
+build, bpftrace and its dependencies can be built in about 30 minutes.
+
+Subsequent builds will not have this overhead, once the embedded dependencies
+are built and cached, and will proceed at the comparable speeds as dynamically
+linked builds.
+
+## CI environments
+
+Internal CI environments with any sort of caching mechanism should be able to
+just cache any external project's build directories, and the overhead of
+building on CI shouldn't be any different than other CI jobs.
+
+# Embedding clang
+
+Clang is the actual dependency of bpftrace, as the src/ast folder pulls it in
+for the interface that bpftraces uses for LLVM's code generation platform, which
+ultimately outputs BPF instructions.
```

Comment:
I don't think this is true, is it?

src/ast/ has llvm deps:
https://github.com/iovisor/bpftrace/blob/master/src/ast/codegen_llvm.h#L11-L14

src/clang_parser.{h,cpp} has clang and llvm deps:
https://github.com/iovisor/bpftrace/blob/master/src/clang_parser.h#L4
https://github.com/iovisor/bpftrace/blob/master/src/clang_parser.cpp#L4

clang is used to parse C headers, then we use llvm to emit bpf bytecode

Reply:
thanks for pointing this out, also TIL